accounts/tests.py

```diff
@@ -1,15 +1,229 @@
+            is_active=True,
+            is_researcher=False,
+            nickname="MySpace Tom",
+        )
+        self.participant.set_password(self.test_password)
+        self.participant.save()
+
+        # Site fixture enabling login
+        self.fake_site = G(Site, id=1)
+
+        # FlatPage fixture enabling login redirect to work.
+        self.home_page = G(FlatPage, url="/")
+        self.home_page.sites.add(self.fake_site)
+        self.home_page.save()
+
+    def test_researcher_registration_flow(self):
+        response = self.client.post(
+            reverse("accounts:researcher-registration"),
+            {
+                "username": "tester@test.com",
+                "password1": self.test_password,
+                "password2": self.test_password,
+                "given_name": "Test",
+                "family_name": "Person",
+                "nickname": "Testman",
+            },
+            follow=True,
+        )
+
+        self.assertEqual(
+            response.redirect_chain, [(reverse("accounts:2fa-setup"), 302)]
+        )
+        self.assertEqual(response.status_code, 200)
+
+        # Mock correctly entered OTP
+        otp = response.context["user"].otp
+        response = self.client.post(
+            reverse("accounts:2fa-setup"), {"otp_code": otp.provider.now()}, follow=True
+        )
+
+        self.assertTrue(response.wsgi_request.session[TWO_FACTOR_AUTH_SESSION_KEY])
+        self.assertEqual(response.redirect_chain, [("/exp/studies/", 302)])
+        self.assertEqual(response.status_code, 200)
+
+    def test_2fa_flow_wrong_otp_reload_page(self):
+        response = self.client.post(
+            reverse("accounts:researcher-registration"),
+            {
+                "username": "another@test.com",
+                "password1": self.test_password,
+                "password2": self.test_password,
+                "given_name": "Second",
+                "family_name": "Person",
+                "nickname": "Testwoman",
+            },
+            follow=True,
+        )
+
+        # Mock incorrectly entered OTP - off by one.
+        otp = response.context["user"].otp
+        correct = otp.provider.now()
+        incorrect = str(int(correct) + 1)
+        old_qr_svg = response.context["svg_qr_code"]
+        response = self.client.post(
+            reverse("accounts:2fa-setup"), {"otp_code": incorrect}, follow=True
+        )
+        new_qr_svg = response.context["svg_qr_code"]
+
+        # Should have reloaded the page with the same QR code
+        self.assertNotIn(TWO_FACTOR_AUTH_SESSION_KEY, response.wsgi_request.session)
+        self.assertEqual(old_qr_svg, new_qr_svg)
+
+    def test_researcher_login_redirect_to_2FA_verification(self):
+        response = self.client.post(
+            reverse("login"),
+            {"username": self.researcher_email, "password": self.test_password},
+            follow=True,
+        )
+        self.assertTrue(response.status_code == 200)
```

Comment:
is there a reason not to use assertEqual?

Reply:
nah, not quite sure why I did that - maybe I was switching from testing one thing to another and just switched the tested element without thinking about the semantics? Regardless, it's changed now